Update README.md (#32527)
vespa-engine/vespa: TODO.md

**Difficulty:** Low<br/>
**Skills:** Java

Query profiles make it simple to support multiple buckets, behavior profiles for different use cases, etc. by providing
bundles of parameters accessible to Searchers processing queries. Writes go through a similar chain of processors -

> Document Processors, but have no equivalent support for parametrization. This is to allow the configuration of document

Reviewer: ```suggestion
Document Processors, but have no equivalent support for parametrization. This is to allow configuration of document
```

**Difficulty:** Low<br/>
**Skills:** Java

There is currently support for creating Application instances programmatically in Java to unit-test application package
functionality (see com.yahoo.application.Application). However, only Java component functionality can be tested in this
way as the content layer is not available, being implemented in C++. A Java implementation, of some or all of the

> functionality would enable developers to do more testing locally within their IDE. This is a medium effort because

Reviewer: ```suggestion
functionality would enable developers to do more testing locally within their IDE. This is medium effort because
```

**Skills:** C++, Java, distributed systems, performance, multithreading, network, distributed consistency

Vespa instances distribute data automatically within clusters, but these clusters are meant to consist of co-located
machines - the distribution algorithm is not suitable for global distribution across data centers because it cannot
seamlessly tolerate data center-wide outages and does not attempt to minimize bandwidth usage between data centers.
The application usually achieves global presence instead by setting up multiple independent instances in different
data centers and write to all in parallel. This is robust and works well on average, but puts an additional burden on
applications to achieve cross-datacenter data consistency on datacenter failures, and does not enable automatic

> data recovery across datacenter, such that data redundancy is effectively required within each datacenter.

Reviewer: ```suggestion
data recovery across data centers, such that data redundancy is effectively required within each data center.
```

large tensors (which barely fit in memory) and/or dynamically changing tensors (online learning of global models).
These use cases require support for global tensors (tensors available locally on all content nodes during execution
but not sent with the query or residing in documents) which are not configured as part of the application package but
which are written independently and dynamically update-able at a high write rate. To support this at a large scale, with a
high write rate, we need a small cluster of nodes storing the source of truth of the global tensor which has
perfect consistency. This in turn must push updates to all content nodes in a best-effort fashion given a fixed bandwidth

> budget, such that query execution and document write traffic are prioritized over ensuring perfect consistency of global

Reviewer: ```suggestion
budget, such that query execution and document write traffic is prioritized over ensuring perfect consistency of global
```

**Difficulty:** Low<br/>
**Skills:** Knowledge of a decent HTTP/2 library in some language

/document/v1 is a RESTified HTTP API that exposes the Vespa Document API to the
outside of the application's Java containers. The design of this API is simple,
with each operation modeled as a single HTTP request, and its result as
a single HTTP response. While it was previously not possible to achieve comparable

> throughput using this API to what the undocumented, custom-protocol /feed API offered,

Reviewer: ```suggestion
throughput using this API to what the undocumented, custom-protocol /feedapi offered,
```
